File: src/main/java/com/example/movieapi/web/rest/RatingController.java
package com.example.movieapi.web.rest;


import com.example.movieapi.chat.model.Message;
import com.example.movieapi.configuration.RedisPublisher;
import com.example.movieapi.configuration.RedisSubscriber;
import com.example.movieapi.dto.ResponseRatingDTO;
import com.example.movieapi.entity.Rating;
import com.example.movieapi.repository.RatingRepository;
import com.example.movieapi.service.RatingLiveService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping("/rating")
public class RatingController {

    @Autowired
    private RedisPublisher messagePublisher;

    @Autowired
    private RedisTemplate<String, Object> redisTemplate;

    @Autowired
    private RatingLiveService ratingLiveService;


    @Autowired
    private RatingRepository ratingRepository;

    //평점 메김
    //@PostMapping("/update")
    public void publish( @RequestBody Rating rating) {
        messagePublisher.publish(rating.toString());
    }
    //평점 리스트
    @GetMapping("/list")
    public List<String> getMessages(){
        return RedisSubscriber.messageList;
    }

    //db에서 저장되는 값이 들옴
    @GetMapping("/list/{movieId}")
    public ResponseEntity<List<ResponseRatingDTO>> getRatingList(@PathVariable Long movieId) {
        List<ResponseRatingDTO> ratingList = ratingLiveService.getRating(movieId);
        return ResponseEntity.ok(ratingList);
    }


    //db에 저장,
    @PostMapping("/update")
    public ResponseEntity<Rating> rateMovie(@RequestBody Rating rating) {
        Optional<Rating> existingRating = ratingRepository.findByUserIdAndMovieId(rating.getUserId(), rating.getMovieId());
        if (existingRating.isPresent()) {
            Rating updatedRating = existingRating.get();
            updatedRating.setRating(rating.getRating());
            ratingRepository.save(updatedRating);
            return ResponseEntity.ok(updatedRating);
        } else {
            Rating newRating = ratingRepository.save(rating);
            return ResponseEntity.ok(newRating);
        }
    }

    @GetMapping("/{userId}/{movieId}")
    public ResponseEntity<Rating> getRating(@PathVariable Long userId, @PathVariable Long movieId) {
        Optional<Rating> rating = ratingRepository.findByUserIdAndMovieId(userId, movieId);
        return rating.map(ResponseEntity::ok).orElseGet(() -> ResponseEntity.noContent().build());
    }
}

File: src/main/java/com/example/movieapi/web/rest/RatingController.kt
package com.example.movieapi.web.rest

import com.example.movieapi.dto.RatingDTO
import com.example.movieapi.service.RatingService
import com.example.movieapi.service.RedisRatingService
import com.example.movieapi.utils.mapper.RatingMapper
import lombok.extern.log4j.Log4j2
import org.apache.logging.log4j.LogManager
import org.springframework.http.HttpStatus
import org.springframework.http.HttpStatusCode
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@Log4j2
@RestController
@RequestMapping("/rating")
class RatingController (
    private val ratingRedisService: RedisRatingService,
    private val ratingService: RatingService,
    private val ratingMapper: RatingMapper
) {

    val logger = LogManager.getLogger(RatingController::class.java)!!

//    @PostMapping("/add")
//    fun addRating(@RequestBody ratingDTO: RatingDTO) : ResponseEntity<RatingDTO> {
//        return ResponseEntity(ratingService.createRating(ratingDTO), HttpStatus.OK)
//    }
//
//    @PostMapping("/save")
//    fun saveRating(@RequestBody ratingDTO: RatingDTO): ResponseEntity<Any> {
//        val rating = ratingMapper.toEntity(ratingDTO)
//        logger.info(rating.toString())
//        ratingRedisService.saveRating(rating)
//        return ResponseEntity.ok().build()
//    }
//
//    @GetMapping("/{id}")
//    fun findRatingById(@PathVariable id: Long): ResponseEntity<Any> {
//        val rating = ratingRedisService.findRatingById(id)
//        return if (rating != null) {
//            ResponseEntity.ok(rating)
//        } else {
//            ResponseEntity.status(HttpStatus.NOT_FOUND).body("Rating not found")
//        }
//    }
}